cds.middlewares.after (#968)
cap-js/docs: node.js/cds-serve.md

For each service served at a certain protocol, the framework registers a configurable set of express middlewares by default like so:

```js
app.use (cds.middlewares.before, protocol_adapter, cds.middlewares.after)
```

The standard set of middlewares uses the following order:

```js
cds.middlewares.before = [
  context(),   // provides cds.context
  trace(),     // provides detailed trace logs when DEBUG=trace

>   auth(),      // provides req.user & tenant

Author: ```suggestion
  auth(),      // provides cds.context.user & cds.context.tenant
```

Reviewer: Should I commit this or is anyone else meant to review it?

Author: → https://github.com/cap-js/docs/pull/1094

>   ctx_auth(),  // propagates auth results to cds.context

Author: ```suggestion
```

Author: → https://github.com/cap-js/docs/pull/1094

  ctx_model(), // fills in cds.context.model
]
```

```js
cds.middlewares.after = [
  cds_error_handler(), // provides final error handling
]
```

> ::: tip
> In order for a custom error middleware to be invoked, it must be registered _before_ the built-in `cds_error_handler`. That is, it must be added, for example, via `cds.middlewares.after.unshift()`.
> :::

Reviewer: ```suggestion
::: tip Custom error middleware before `cds_error_handler`
To invoke a custom error middleware successfully, you must register it _before_ the built-in `cds_error_handler`. You can achieve that, for example, by adding the middleware using `cds.middlewares.after.unshift()`.
:::
```

Be aware that using an absolute path will disallow serving the service at multiple protocols.

### PATCH vs. PUT vs. Replace

The HTTP method `PATCH` is meant for partial modification of an _existing resource_.
`PUT`, on the other hand, is meant for ensuring a resource exists
That is, if it doesn't yet exists, it gets created.
If it does exist, it gets updated to reflect the request's content.

This content, however, may be incomplete.
By default, the values for not listed keys are not touched.
The rationale being that default values are known and clients have the option to send full representations, if necessary.

The following table shows the Node.js runtime's configuration options and their respective default value:

| Flag                                         | Behavior                                 | Default |
|----------------------------------------------|------------------------------------------|---------|

> | <Config>cds.runtime.patch_as_upsert</Config> | Create resource if it does not yet exist | false   |
> | <Config>cds.runtime.put_as_upsert</Config>   | Create resource if it does not yet exist | true    |
> | <Config>cds.runtime.put_as_replace</Config>  | Payload is enriched with default values  | false   |

Reviewer: ```suggestion
| <Config keyOnly>cds.runtime.patch_as_upsert = false</Config> | Create resource if it does not yet exist   | false   |
| <Config keyOnly>cds.runtime.put_as_upsert = true</Config>     | Create resource if it does not yet exist    | true    |
| <Config keyOnly>cds.runtime.put_as_replace = false</Config>   | Payload is enriched with default values  | false   |
```
